### DISCUSSION TITLE
Data messages delayed until the app is brought to the foreground

### DISCUSSION THREAD

**Discussion by anhnch** [invertase/react-native-firebase#6391] [Q&A] [answered]
I having a problem with receiving notification messages. I use cloud function to push data messages to the client and use Notifee to display it. Everything is great, I can send and receive in foreground, background and quit state. 

```typescript
// index.js
const backgroundHandler = (message) => {
    console.log('Message comes when app is in the background/quit state');

    // ... check some condition before deciding to display the message or not
    if (!shouldDisplay) return; // suppress the message

    notifee.displayNotification({
        title: message.data.title,
        body: message.data.body
    })
}
n12.setBackgroundMessageHandler(backgroundHandler);
```
```typescript
// App.tsx
const foregroundHandler = (message) => {
     console.log('Message comes when app is in the foreground');
}
useEffect(() => {
    messaging().onMessage(foregroundHandler);
}, [])
```

Everything is fine, very straightforward. Then one day one of my iphone (iOS 15.5) doesn't show message when the app is in background/quit state anymore. The message did come to the device, but somehow it waits/delays until the app comes into the foreground state.

The backgroundHandler should be triggered if a message comes when the app is in background/quit state. But it's never called. When I open the app, the foregroundHandler is triggered right after that.

Same code, but the other iphone (iOS 14.7) works fine.

* I wonder if somehow firebase (or iOS) detects the sending pattern and reduce the priority of the messages due to too many testing messages?
* Or maybe it "thinks" that messages that are not opened by users will become less important?
* Or data messages are less important than notification messages?
* What, Firebase or iOS, may cause this behavior?
* Or something new in iOS 15.5?

Have you ever had this situation?

**Comment by anhnch** [comment]
When I test on different phones with different version of iOS I can narrow down
- For iphone 11 pro max iOS 15: message delayed
- For iPhone Xs Max iOS 14.7: OK
- For iPhone 7 Plus iOS 15.5: message delayed
- For iPad mini iOS 15.5: OK

What I can say is that iOS 15.5 alone is not a problem, but also the device. It seems that the message is only delayed on iPhone wth iOS 15.5 AND it's a data message. Notification message is Ok. 

I've tried changing parameters when sending messages, but it looks like it's a client side problem.
That's for now. Need more testings

**Comment by anhnch** [comment, marked as answer]
I found the reason why! It's not the iOS, it's not firebase, it's not the device
```
We must go to iOS settings > General > Background App Refresh and set it to ON
```
The message did come, but if the app is not allowed to refresh itself at the background, the listener is not triggered. The message is delayed until it comes to the foreground.

This took me 3 days scratching head crazily. Now when I search with the keyword "firebase messaging background app refresh" I found https://rnfirebase.io/messaging/usage#ios-background-limitation it states very clearly. When I said "one day, it doesn't work anymore". It's because the iPhone must have gone to the Low Power Mode, the Background App Refresh is turned off automatically.

How did I miss this out!

**Reply by mikehardy** [reply]
I think a community member just added that as a docs PR quite recently, perhaps after you started noticing it. Data only messages are very subtle ...

**Reply by mikehardy** [reply]
Also worth noting i have a PR for the net info module (different repo - the community module) that adds background refresh status, I'm not sure if I posted it or merged it there yet though? And I think notifee.app has an API for getting the status as well, in case you want to educate users then send them to settings for some app feature to improve delivery

**Reply by anhnch** [reply]
> Also worth noting i have a PR for the net info module (different repo - the community module) that adds background refresh status, I'm not sure if I posted it or merged it there yet though? And I think notifee.app has an API for getting the status as well, in case you want to educate users then send them to settings for some app feature to improve delivery

Thanks for the advice. Yes, I'm gonna check and remind the user.

**Reply by mikehardy** [reply]
Ah - I just saw that I posted it, but it's out of date and I did not merge yet.
Here's the basic of how to do it though, I'm actually using this in production (via patch-package, just not updated quite yet to newest netinfo release and needs some tiny change...). So the code should work.
Or notifee.app of course

https://github.com/react-native-netinfo/react-native-netinfo/pull/523

**Reply by anhnch** [reply]
Yes I saw your PR and merge pending status. Do you have plan to implement it in notifee? For convenience, you know.
